v24.1.0-alpha.4 release notes (#18415)
cockroachdb/docs: src/current/_includes/releases/v24.1/v24.1.0-alpha.4.md

## v24.1.0-alpha.4

Release Date: March 25, 2024

{% include releases/release-downloads-docker-image.md release=include.release %}

<h3 id="v24-1-0-alpha-4-security-updates">Security updates</h3>

- When [configuring logs](../v24.1/configure-logs.html#output-to-files), `file-permissions` are now applied literally, such that `file-permissions: 644` will result in files with permissions matching `644` (instead of the previous behavior's `640`). Previously, CockroachDB's `umask` (which is always at least `007`) was being applied after the `file-permissions` field was used to create files, meaning the resulting permissions did not match those specified in the log configuration. [#120669][#120669]

<h3 id="v24-1-0-alpha-4-general-changes">General changes</h3>

> - The following [metrics](../v24.1/metrics) were added for observability of per-store disk events: 

Reviewer: ```suggestion
- The following [metrics](../v24.1/metrics.html) were added for observability of per-store disk events: 
```

<h3 id="v24-1-0-alpha-4-{{-site.data.products.enterprise-}}-edition-changes">{{ site.data.products.enterprise }} edition changes</h3>

- `server.controller.default_target_cluster` can now be set to any virtual cluster name by default, including a virtual cluster yet to be created or have service started. [#120080][#120080]
- The [`READ COMMITTED`](../v24.1/read-committed.html) isolation level now requires the cluster to have a valid enterprise license. [#120154][#120154]
- The new boolean changefeed option `ignore_disable_changefeed_replication`, when set to `true`, prevents the changefeed from filtering events even if CDC filtering is configured via the `disable_changefeed_replication` [session variable](../v24.1/session-variables.html), `sql.ttl.changefeed_replication.disabled` [cluster setting](../v24.1/cluster-settings.html), or the `ttl_disable_changefeed_replication` [table storage parameter](../v24.1/alter-table.html#table-storage-parameters). [#120255][#120255]

<h3 id="v24-1-0-alpha-4-sql-language-changes">SQL language changes</h3>

- Added support for the [PL/pgSQL](../v24.1/plpgsql.html) `COMMIT` and `ROLLBACK` statements. [#119647][#119647]

> - [Identity columns](..v24.1/create-table#identity-columns) now support enhanced sequence management through the [`ALTER [COLUMN] column_name SET sequence_option`](../v24.1/alter-table.html#alter-column) and [`ALTER [COLUMN] column_name RESTART [WITH restart]`](../v24.1/alter-table.html#alter-column) commands. This update facilitates the fine-tuning of identity column sequences. [#119432][#119432]

Reviewer: ```suggestion
- [Identity columns](..v24.1/create-table#identity-columns.html) now support enhanced sequence management through the [`ALTER [COLUMN] column_name SET sequence_option`](../v24.1/alter-table.html#alter-column) and [`ALTER [COLUMN] column_name RESTART [WITH restart]`](../v24.1/alter-table.html#alter-column) commands. This update facilitates the fine-tuning of identity column sequences. [#119432][#119432]
```

- The DB Console **SQL Activity** [**Statement Fingerprint**](../v24.1/ui-statements-page.html#statement-fingerprint-page) page has replaced the **Failed?** boolean column with a **Failure Count** column that shows the number of failed executions for the given statement fingerprint.<br /> In the **SQL Activity** table, the same statement fingeprint no longer appears in separate rows for failed executions and successful executions. Instead, they are combined into a single statement fingerprint. [#120236][#120236]
- The DB Console now displays an alert message when a license has expired or will expire in fewer than 15 days. [#120490][#120490]

<h3 id="v24-1-0-alpha-4-bug-fixes">Bug fixes</h3>

- Fixed a bug with [`DROP SCHEMA ... CASCADE`](../v24.1/drop-schema.html) that could lead to dangling function references in other schemas accessing any functions. [#119932][#119932]
- Fixed a bug where a [`RESTORE`](../v24.1/restore.html) of a backup that itself contained a table created by the `RESTORE` of a table with an in-progress [`IMPORT INTO`](../v24.1/import-into.html) would fail to restore all rows. [#120414][#120414]

> - Fixed a bug where [identity columns](..v24.1/create-table#identity-columns) without any configured sequence options did not display the default values for identity attributes in `information_schema`. [#119459][#119459]

Reviewer: ```suggestion
- Fixed a bug where [identity columns](..v24.1/create-table#identity-columns.html) without any configured sequence options did not display the default values for identity attributes in `information_schema`. [#119459][#119459]
```

- A new boolean changefeed option named `ignore_disable_changefeed_replication` has been added. When set, the changefeed will not filter events even if CDC filtering is configured, which is currently available via the `disable_changefeed_replication` [session variable](../v24.1/session-variables.html), `sql.ttl.changefeed_replication.disabled` [cluster setting](../v24.1/cluster-settings.html), and the `ttl_disable_changefeed_replication` [table storage parameter](../v24.1/alter-table.html#table-storage-parameters). [#120255][#120255]

<h3 id="v24-1-0-alpha-4-sql-language-changes">SQL language changes</h3>

- Added support for the [PL/pgSQL](../v24.1/plpgsql.html) `COMMIT` and `ROLLBACK` statements. [#119647][#119647]
- Identity columns now support enhanced sequence management through the [`ALTER [COLUMN] column_name SET sequence_option`](../v24.1/alter-table.html#alter-column) and [`ALTER [COLUMN] column_name RESTART [WITH restart]`](../v24.1/alter-table.html#alter-column) commands. This update facilitates the fine-tuning of identity column sequences. [#119432][#119432]

> - It is now possible to use the `STRICT` option with [`SELECT ... INTO`](../v24.1/plpgsql.html#assign-a-result-to-a-variable) and [`RETURNING ... INTO`](../v24.1/plpgsql.html#assign-a-result-to-a-variable) in order to enforce that a SQL statement within a [PL/pgSQL](../v24.1/plpgsql.html) routine returns exactly one row. [#120486][#120486]

Author: This should actually be PL/pgSQL afaik.

- Added a [session setting](../v24.1/session-variables.html) `plpgsql_use_strict_into`, which causes PL/pgSQL [`SELECT ... INTO`](../v24.1/plpgsql.html#assign-a-result-to-a-variable) and [`RETURNING ... INTO`](../v24.1/plpgsql.html#assign-a-result-to-a-variable) to require exactly one row from the SQL statement, similar to Oracle behavior. [#120486][#120486]

> - Added a new `failure_count INT NOT NULL` column to `crdb_internal.node_statement_statistics`. It represents the number of recorded statement execution failures for the given statement, as a new component of the overall statistics. [#120236][#120236]

Author: This is still a gray area. We document `crdb_internal` but also caveat that many of the tables shouldn't be used in production (including this one). Leaving this in until there's a clearer decision on `crdb_internal`

- The `FORCE_INVERTED_INDEX` hint is now supported. This makes the [optimizer](../v24.1/cost-based-optimizer.html) prefer a query plan scan over any inverted index of the hinted table. The query will result in an error if no such query plan can be generated. [#120384][#120384]
- The `REPAIRCLUSTERMETADATA` privilege now has an alternative name: `REPAIRCLUSTER`. Both names can be used interchangeably. [#116844][#116844]

<h3 id="v24-1-0-alpha-4-operational-changes">Operational changes</h3>

> - The [cluster setting](../v24.1/cluster-settings.html) `admission.wal.failover.unlimited_tokens.enabled` can be set to `true` to cause unlimited admission tokens during write-ahead log (WAL) failover. This should not be changed without significant expertise, since the default, which preserves the token counts from the preceding non-WAL-failover interval, is expected to be safer. [#120135][#120135]

Author: I'm going to leave this as-is because the specific wording was discussed with the engineer.

- Setting the new [`cockroach start`](../v24.1/cockroach-start.html) option `--wal-failover=among-stores` or `COCKROACH_WAL_FAILOVER=among-stores` environment variable will configure a multi-store CockroachDB node to failover a store's write-ahead log (WAL) to another store's data directory. Failing over the write-ahead log may allow some operations against a store to continue to complete despite temporary unavailability of the underlying storage.. [#120509][#120509]
- Introduced a new `storage.wal_failover.unhealthy_op_threshold` [cluster setting](../v24.1/cluster-settings.html) for configuring the latency threshold at which a WAL write is considered unhealthy. [#120509][#120509]
- Added two new metrics to track the status of the SQL activity update job, which is used to pre-aggregate top K information within the SQL stats subsytem and write the results to `system.statement_activity` and `system.transaction_activity`:
	- `sql.stats.activity.updates.successful`: Number of successful updates made by the SQL activity updater job.
	- `sql.stats.activity.update.latency`: The latency of updates made by the SQL activity updater job. Includes failed update attempts. [#120522][#120522]

> - Added a new counter metric, `sql.stats.flush.done_signals.ignored`, that tracks the number of times the SQL activity update job ignored the signal indicating a flush has completed. This may indicate that the SQL activity update job is taking longer than expected to complete. [#120522][#120522]

Reviewer: ```suggestion
- The new counter metric, `sql.stats.flush.done_signals.ignored` tracks the number of times the SQL Activity Update job has ignored the signal that indicates that a flush has completed. Growth of this metric can indicate that the SQL activity Update job is taking longer than expected to complete. [#120522][#120522]
```

> - Added a new counter metric, `sql.stats.activity.updates.failed`, to measure the number of update attempts made by the SQL activity update job that failed with errors. The SQL activity update job is used to pre-aggregate top K information within the SQL stats subsystem and write the results to `system.statement_activity` and `system.transaction_activity`. [#120522][#120522]

Reviewer: ```suggestion
- The new counter metric `sql.stats.activity.updates.failed` measures the number of times update attempts the SQL Activity Update job has attempted which have failed with errors. The SQL activity update job pre-aggregates top K information within the SQL Statistics subsystem and writes the results to `system.statement_activity` and `system.transaction_activity`. [#120522][#120522]
```
Consider combining with the note above, which relates to the same PR and has a redundant definition of the SQL Activity Update job.

Author: Good catch -- I'm just going to cut the second sentence.

> - Added a new counter metric, `sql.stats.flush.fingerprint.count`, that tracks the number of unique statement and transaction fingerprints included in the SQL stats flush. [#120522][#120522]

Reviewer: ```suggestion
- The new counter metric `sql.stats.flush.fingerprint.count` tracks the number of unique statement and transaction fingerprints flushed from the SQL statistics flush. [#120522][#120522]
```

> - Added `node_id`, `dir`, and `wal_failover_path` fields to the `/_status/stores` endpoint. These indicate the store's node ID, data directory, and path to the configured WAL failover secondary, if configured. [#120677][#120677]

Reviewer: ```suggestion
- The `/_status/stores` now includes  `node_id`, `dir`, and `wal_failover_path` fields to show the store's node ID, data directory, and path to the configured WAL failover secondary, if configured. [#120677][#120677]
```

<h3 id="v24-1-0-alpha-4-command-line-changes">Command-line changes</h3>

> - Added a new `--go-gc-percent` flag for the [`cockroach start`](../v24.1/cockroach-start.html) command that controls the garbage collection target percentage on the Go runtime, mirroring the existing GOGC environment variable. A garbage collection is triggered when the ratio of freshly allocated data to live data remaining after the previous collection reaches this percentage. If left unspecified and a Go soft memory limit is configured (i.e., not explicitly disabled via `--max-go-memory` or `GOMEMLIMIT`), the GC target percentage defaults to 300%. If set to a negative value, the flag disables the target percentage garbage collection heuristic, leaving only the soft memory limit heuristic to trigger garbage collection. [#119605][#119605]

Reviewer: ```suggestion
- The new `--go-gc-percent` flag of the [`cockroach start`](../v24.1/cockroach-start.html) command controls the garbage collection target percentage of the Go runtime, mirroring the existing `GOGC` environment variable. A garbage collection is triggered when the ratio of freshly allocated data to live data remaining after the previous collection reaches this percentage. If left unspecified and if a Go soft memory limit is configured (i.e., not explicitly disabled via `--max-go-memory` or `GOMEMLIMIT`), the garbage collection target percentage defaults to 300%. Setting the flag to a negative value disables  the target percentage garbage collection heuristic, and only the soft memory limit heuristic triggers garbage collection. [#119605][#119605]
```

<h3 id="v24-1-0-alpha-4-db-console-changes">DB Console changes</h3>

> - The [**Queues** dashboard](../v24.1/ui-queues-dashboard.html) will now display lease queue metrics, similar to the other replica queues. [#119386][#119386]

Reviewer: ```suggestion
- The [**Queues** dashboard](../v24.1/ui-queues-dashboard.html) now includes lease queue metrics. [#119386][#119386]
```
"other replica queues" is confusing. I suspect this is what it means.

> - In the DB Console's **SQL Activity** page, the [**Statement Fingerprint**](../v24.1/ui-statements-page.html#statement-fingerprint-page) page now displays "Failure Count" in place of "Failed?", showing the number of failed executions for the given statement fingerprint instead of a "Yes"/"No" value. In addition, previously in the **SQL Activity** table the same statement fingeprint would appear twice: once for failed executions, and once for successful executions. These are no longer considered to be separate rows. Instead, they are combined into a single statement fingerprint. [#120236][#120236]

Reviewer: ```suggestion
- The DB Console **SQL Activity** [**Statement Fingerprint**](../v24.1/ui-statements-page.html#statement-fingerprint-page) page has replaced the **Failed** boolean column with a  "Failure Count" column that shows the number of failed executions for the given statement fingerprint.<br /> With this change, in the **SQL Activity** table, the same statement fingeprint no longer appears in separate rows for failed executions and successful executions.  Instead, they are combined into a single statement fingerprint. [#120236][#120236]
```

> - An alert message is now displayed in the DB Console when a license is expired or there are fewer than 15 days left before it expires. [#120490][#120490]

Reviewer: ```suggestion
- The DB Console now displays an alert message when a license has expired or will expire in fewer than 15 days. [#120490][#120490]
```

<h3 id="v24-1-0-alpha-4-bug-fixes">Bug fixes</h3>

> - Previously, [`DROP SCHEMA ... CASCADE`](../v24.1/drop-schema.html) could lead to dangling function references in other schemas accessing any functions. This is now fixed. [#119932][#119932]

Reviewer: ```suggestion
- Fixed a bug [`DROP SCHEMA ... CASCADE`](../v24.1/drop-schema.html) that could lead to dangling function references in other schemas accessing any functions. [#119932][#119932]
```

> - Fixed an issue in which a [`RESTORE`](../v24.1/restore.html) of a backup that itself contained a table created by the `RESTORE` of a table with an in-progress [`IMPORT INTO`](../v24.1/import-into.html) would fail to restore all rows. [#120414][#120414]

Reviewer: ```suggestion
- Fixed a bug where a [`RESTORE`](../v24.1/restore.html) of a backup that itself contained a table created by the `RESTORE` of a table with an in-progress [`IMPORT INTO`](../v24.1/import-into.html) would fail to restore all rows. [#120414][#120414]
```

> - Identity columns without specified sequence options now correctly show default values for identity attributes in `information_schema`. [#119459][#119459]

Reviewer: ```suggestion
- Fixed a bug where [identity column](..v24.1/create-table#identity-columns) without any configured sequence options did not display the default values for identity attributes in `information_schema`. [#119459][#119459]
```

> - Previously, [`GRANT ... ON ALL TABLES`](../v24.1/grant.html) could fail if sequences existed and they did not support a privilege (e.g., `BACKUP`). This is now fixed. [#120685][#120685]

Reviewer: ```suggestion
- Fixed a bug where a [`GRANT ... ON ALL TABLES`](../v24.1/grant.html) statement could fail if sequences existed and they did not support a privilege (e.g., `BACKUP`). [#120685][#120685]
```

> - Previously, [`EXPLAIN (DDL)`](../v24.1/explain.html) would generate event log entries for schema changes even though they weren't executed. This is now fixed. [#120563][#120563]

Reviewer: ```suggestion
- Fixed a bug where an [`EXPLAIN (DDL)`](../v24.1/explain.html) statement would generate event log entries for schema changes that were not executed.  [#120563][#120563]
```
